[Crash] TypeCastException in FingerprintManager

Reported in Play Store pre-launch checklist tests on 
Xperia XZ1 Compact with Android 8.0:


>Issue: java.lang.RuntimeException: Unable to create application mozilla.lockbox.LockboxApplication: kotlin.TypeCastException: null cannot be cast to non-null type android.hardware.fingerprint.FingerprintManager

```
FATAL EXCEPTION: main
Process: mozilla.lockbox, PID: 13148
java.lang.RuntimeException: Unable to create application mozilla.lockbox.LockboxApplication: kotlin.TypeCastException: null cannot be cast to non-null type android.hardware.fingerprint.FingerprintManager
	at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5825)
	at android.app.ActivityThread.-wrap1(Unknown Source:0)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1666)
	at android.os.Handler.dispatchMessage(Handler.java:105)
	at android.os.Looper.loop(Looper.java:251)
	at android.app.ActivityThread.main(ActivityThread.java:6572)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:240)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:767)
Caused by: kotlin.TypeCastException: null cannot be cast to non-null type android.hardware.fingerprint.FingerprintManager
	at mozilla.lockbox.LockboxApplication.onCreate(LockboxApplication.kt:59)
	at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1119)
	at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5822)
	... 8 more
```

_Originally posted by @devinreams in https://github.com/mozilla-lockbox/lockbox-android/issues/194#issuecomment-437963871_

--- a/app/src/main/java/mozilla/lockbox/store/FingerprintStore.kt
+++ b/app/src/main/java/mozilla/lockbox/store/FingerprintStore.kt
@@ -37,7 +37,7 @@ open class FingerprintStore(
     val dispatcher: Dispatcher = Dispatcher.shared
 ) : ContextStore {
     internal val compositeDisposable = CompositeDisposable()
-    open lateinit var fingerprintManager: FingerprintManager
+    open var fingerprintManager: FingerprintManager? = null
     open lateinit var keyguardManager: KeyguardManager
     private lateinit var authenticationCallback: AuthenticationCallback
 
@@ -75,7 +75,7 @@ open class FingerprintStore(
     }
 
     override fun injectContext(context: Context) {
-        fingerprintManager = context.getSystemService(Context.FINGERPRINT_SERVICE) as FingerprintManager
+        fingerprintManager = context.getSystemService(Context.FINGERPRINT_SERVICE) as? FingerprintManager
         keyguardManager = context.getSystemService(Context.KEYGUARD_SERVICE) as KeyguardManager
         authenticationCallback = AuthenticationCallback(context)
     }
@@ -84,7 +84,7 @@ open class FingerprintStore(
         get() = isFingerprintAuthAvailable || isKeyguardDeviceSecure
 
     open val isFingerprintAuthAvailable: Boolean
-        get() = fingerprintManager.isHardwareDetected && fingerprintManager.hasEnrolledFingerprints()
+        get() = (fingerprintManager?.isHardwareDetected ?: false) && (fingerprintManager?.hasEnrolledFingerprints() ?: false)
 
     open val isKeyguardDeviceSecure get() = keyguardManager.isDeviceSecure
 
@@ -103,7 +103,7 @@ open class FingerprintStore(
         }
         cancellationSignal = CancellationSignal()
         selfCancelled = false
-        fingerprintManager.authenticate(cryptoObject, cancellationSignal, 0, authenticationCallback, null)
+        fingerprintManager?.authenticate(cryptoObject, cancellationSignal, 0, authenticationCallback, null)
     }
 
     private fun stopListening() {

--- a/app/src/test/java/mozilla/lockbox/presenter/LockedPresenterTest.kt
+++ b/app/src/test/java/mozilla/lockbox/presenter/LockedPresenterTest.kt
@@ -37,8 +37,8 @@ class LockedPresenterTest {
     }
 
     open class FakeFingerprintStore : FingerprintStore() {
-        override var fingerprintManager: FingerprintManager =
-            RuntimeEnvironment.application.getSystemService(Context.FINGERPRINT_SERVICE) as FingerprintManager
+        override var fingerprintManager: FingerprintManager? =
+            RuntimeEnvironment.application.getSystemService(Context.FINGERPRINT_SERVICE) as? FingerprintManager
 
         override var keyguardManager: KeyguardManager =
             spy(RuntimeEnvironment.application.getSystemService(Context.KEYGUARD_SERVICE) as KeyguardManager)

